tools/clang/lib/SPIRV/EmitSPIRVAction.cpp

```diff
@@ -58,14 +99,34 @@ class SPIRVEmitter : public ASTConsumer {
     const uint32_t funcType = translateFunctionType(decl);
     const uint32_t retType = translateType(decl->getReturnType());
 
-    theBuilder.beginFunction(funcType, retType);
+    const uint32_t funcId = theBuilder.beginFunction(funcType, retType);
     // TODO: handle function parameters
     // TODO: handle function body
     const uint32_t entryLabel = theBuilder.bbCreate();
     theBuilder.bbReturn(entryLabel);
     theBuilder.endFunction();
-  }
 
+    // Add an entry point to the module if necessary
+    const std::string hlslEntryFn =
+        theCompilerInstance.getCodeGenOpts().HLSLEntryFunction;
+    if (hlslEntryFn == decl->getNameInfo().getAsString()) {
+      const spv::ExecutionModel em = getSpirvShaderKindFromHlslProfile(
+          theCompilerInstance.getCodeGenOpts().HLSLProfile.c_str());
+      // TODO: Pass correct input/output interfaces to addEntryPoint.
+      theBuilder.addEntryPoint(em, funcId, hlslEntryFn, {});
+
+      // OpExecutionMode declares an execution mode for an entry point.
+      std::vector<uint32_t> execModeInstr;
+      spirv::InstBuilder ib([&execModeInstr](std::vector<uint32_t> &&words) {
+        execModeInstr = std::move(words);
+      });
+      // TODO: Implement the logic to determine the proper Execution Mode based
+      // on Shader Stage and other semantics. (currently using OriginUpperLeft
```

Comment:
`OriginUpperLeft` should be the default of pixel shader only. So we need to check the shader stage here.